haystack/components/agents/agent.py

```diff
@@ -33,6 +34,19 @@
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class _ExecutionContext:
+    """Encapsulates the execution state and configuration"""
+
+    state: State
+    streaming_callback: Optional[StreamingCallbackT]
+    component_visits: dict
+    chat_generator_inputs: dict
+    tool_invoker_inputs: dict
+    counter: int = 0
+    skip_chat_generator: bool = False
```

Comment:
I've introduced a new private dataclass called ExecutionContext (open to a different name) to help with understanding how the Agent can be restarted from a snapshot or fresh.